vivy_encryptor/src/main/java/com/vivy/medicalSticker/v2/EmergencyStickerEncryption.kt

```diff
@@ -0,0 +1,121 @@
+package com.vivy.medicalSticker.v2
+
+import com.vivy.e2e.DecryptionFailed
+import com.vivy.e2e.EncryptionFailed
+import com.vivy.medicalSticker.MedStickerCipherAttr
+import com.vivy.medicalSticker.MedStickerCipherAttr.Companion.CHARLIE
+import com.vivy.medicalSticker.MedStickerEncryption
+import com.vivy.medicalSticker.MedStickerKeyGenerator
+import com.vivy.medicalSticker.common.toHexString
+import com.vivy.medicalSticker.v2.model.EncryptedEmergencySticker
+import com.vivy.support.SecureRandomGenerator
+import com.vivy.symmetric.AesGcmNoPadding
+
+object EmergencyStickerEncryption {
+    var debug: Boolean = false
+    internal const val FIRST_SALT = "5f1288159017d636c13c1c1b2835b8a871780bc2"
+    internal const val CPU_COST = 16384
+    internal const val MEMORY_COST = 10
+    internal const val PARALLELIZATION_PARAM = 1
+    internal const val HASH_LENGTH = 64
+    const val AES_IV_LENGTH = 16 //Intended length of aes key.
+
+    private val gcmNoPadding = AesGcmNoPadding()
+
+    fun encrypt(
+        data: ByteArray,
+        key: ByteArray,
+        iv: ByteArray
+    ):ByteArray{
+        try {
+            return gcmNoPadding.encrypt(data, key, iv)
+        }catch (e: Exception) {
+            throw EncryptionFailed(if (MedStickerEncryption.debug) e else null)
+        }
+    }
+
+    fun encrypt(
+        pin: String,
+        backEndSecret: String,
+        secondSalt:String,
+        data: ByteArray,
+        iv: ByteArray
+    ): EncryptedEmergencySticker {
+
+        val keyPairs = getPinFingerprint(pin, backEndSecret, secondSalt)
+        val encryptedData = encrypt(data, keyPairs.key, iv)
+        val attr = MedStickerCipherAttr(pin.toByteArray(), iv, CHARLIE)
+
+        return EncryptedEmergencySticker(encryptedData, keyPairs.fingerprintFile, attr)
+    }
+
+    /**
+     * pin: From QR Code
+     * pinSalt: Generated from Backend
+     * pinSecret: Generated from Backend
+     */
+
+    internal fun getPinFingerprint(
+        secret: String,
+        backEndSecret: String,
+        secondSalt:String
+    ): EmergencyStickerKeyPairs{
+        val hash = getHash(secret + backEndSecret, secondSalt)
+        val key = hash.dropLast(HASH_LENGTH / 2).toByteArray()
+        val fingerPrintFile = CHARLIE + ":" + hash.drop(HASH_LENGTH / 2).toByteArray().toHexString()
```

Comment:
```suggestion
        val fingerprintFile = CHARLIE + ":" + hash.drop(HASH_LENGTH / 2).toByteArray().toHexString()
```
fingerprint is a single word